File: CarriageNode.py

from node import Node
from line import Line
from junctions import Standard_Points, Buffer

class State():
    def __init__(self, fractional_position, line_occupied, direction):
        self.fractional_position = fractional_position
        self.line_occupied = line_occupied
        self.direction = direction
    
    def as_dict(self):
        return {"fractional_position":self.fractional_position, "line_occupied":self.line_occupied, "direction":self.direction} # Position one step prior to current step
    
    @classmethod
    def from_dict(cls, state_dict):
        return cls(state_dict["fractional_position"], state_dict["line_occupied"], state_dict["direction"])
    
    @classmethod
    def from_carriage_node(cls, node):
        return cls(node.fractional_position, node.line_occupied, node.direction)
    
    def __str__(self):
        return f"{self.fractional_position} on line {self.line_occupied}, Direction: {self.direction}"
    
class Carriage_Node():
    def __init__(self, line_occupied, fractional_position=0):
        self.line_occupied = line_occupied
        self.fractional_position = fractional_position
        self.direction = 1 # 1 if the line segment is start - finish ()
        self.is_ends_node = False # Whether this node is a node at either end of the train. Set to True on train construction
        
        self.save_state()
    
    def save_state(self):
        self.previous_state = State.from_carriage_node(self)
        
    def get_state(self):
        return State.from_carriage_node(self).as_dict()
    
    def get_previous_state(self):
        return self.previous_state.as_dict()
    
    def undo_step(self):
        self.fractional_position = self.previous_state.fractional_position
        self.line_occupied = self.previous_state.line_occupied
        self.direction = self.previous_state.direction
    
    def step(self, velocity, save=False):
        """ Move forward a fixed distance. Returns True on any movement. """
        # Error codes: 0 - success, 1 - failure
        if save:
            self.save_state()
        
        velocity *= self.direction # The velocity in the direction in which the train is treating this segment of track
        line_length = self.line_occupied.length
        fractional_change = velocity / line_length
        new_fractional_position = self.fractional_position + fractional_change
        
        if new_fractional_position >= 0 and new_fractional_position <= 1:
            # Staying on same line
            self.fractional_position = new_fractional_position
            return 0
        else:
            if new_fractional_position < 0:
                #print(f"Hit start junction {str(self.line_occupied.start_junction)}")
                junction = self.line_occupied.start_junction
                overshoot = 0 - new_fractional_position
            else: # new_fractional_position > 1
                #print(f"Hit end junction {str(self.line_occupied.end_junction)}")
                junction = self.line_occupied.end_junction
                overshoot = new_fractional_position - 1
            
        # Switch onto new track:
        if type(junction) != Standard_Points:
            return 1 # Train terminates

        nextline = junction.next(self.line_occupied)
        if nextline == None: # Train can't go this way!
            return 1
            
        print(f"Node at {self.get_location()} steps over junction {junction}. Velocity is currently {velocity} and direction is {self.direction}")
        if self.is_ends_node:
            junction.toggle_occupied()
        overshoot = overshoot * line_length # Gives overshoot as an absolute distance
        fractional_overshoot = overshoot / nextline.length # Overshoot as a fraction of the new line

        nextline_starts_here = (nextline.start_junction == junction)
        thisline_starts_here = (self.line_occupied.start_junction == junction)

        if nextline_starts_here:
            self.previous_state.fractional_position = 0
            self.fractional_position = 0 + fractional_overshoot
        else:
            self.previous_state.fractional_position = 1
            self.fractional_position = 1 - fractional_overshoot

        # If tracks are either nose-nose or tail-tail flip direction, otherwise do nothing:
        if (nextline_starts_here and thisline_starts_here) or ((not nextline_starts_here) and (not thisline_starts_here)):
            self.direction *= -1


        self.line_occupied = nextline
        return 0

    def follow(self, ahead_carriage_node, carriage_length, train_velocity):
        # Error Codes: 0 - success, 1 - failure
        relative_train_velocity = train_velocity # Initial value, independent of direction
        train_velocity *= self.direction # The velocity in the direction in which the train is treating this segment of track (absolute velocity on segment)
        # Find all points which are a carriage length away from
        # ahead_carriage_node on this line
        intersections = self.line_occupied.intersects_circle(ahead_carriage_node.get_position(), carriage_length)

        if train_velocity == 0:
            return 0

        def ahead_of(carriage_node1, carriage_node2):
            if train_velocity > 0: # Positive velocity along this line
                return (carriage_node1 > carriage_node2)
            else: # Negative velocity along this line
                return (carriage_node1 < carriage_node2)

        candidate_positions = []

        # Filter out positions "behind" the current position, or ahead of ahead_carriage_node
        for intersection in intersections:
            # Check intersection is ahead of current position
            if ahead_of(intersection,self.fractional_position):
                if self.line_occupied == ahead_carriage_node.line_occupied:
                    # Then the next step must be behind ahead_carriage_node
                    if not ahead_of(intersection, ahead_carriage_node.fractional_position):
                        candidate_positions.append(intersection)
                else:
                    # Not on the same lines, so relative fractional_positions don't matter
                    candidate_positions.append(intersection)

        if len(candidate_positions) == 1:
            # Only one place to move to
            self.fractional_position = candidate_positions[0]
        elif len(candidate_positions) == 0:
            if self.line_occupied == ahead_carriage_node.line_occupied:
                # On same track as node in front, but nowhere to go, so can't move
                return 1
            else:
                # Nowhere to go on this track, but there may be somewhere on next track
                if train_velocity > 0:
                    # Put this node just beyond junction
                    self.fractional_position =  1.0001
                else:
                    # Put this node just beyond junction
                    self.fractional_position = -0.0001
                initial_line_occupied = self.line_occupied
                self.step(0) # Stepping will now switch this node onto the next line
                if self.line_occupied == initial_line_occupied:
                    raise SystemError(f"No valid position for following node at {self.get_location()}")
                else:
                    # Try following ahead_carriage_node, now this node is on the next line
                    self.follow(ahead_carriage_node, carriage_length, relative_train_velocity)
        else:
            raise SystemError(f"More than 1 position available to move following node at {self.get_location()}: {candidate_positions}")
        return 0
        
    def move_to(self, ahead_carriage_node):
        self.line_occupied = ahead_carriage_node.line_occupied
        self.fractional_position = ahead_carriage_node.fractional_position
        self.direction = ahead_carriage_node.direction

    def get_position(self):
        """ Interpolate position of node """
        position = self.line_occupied.interpolate_position(self.fractional_position)
        return position
    
    def get_location(self):
        """ Display position of node """
        return f"{self.get_position()} on line {self.line_occupied}"

File: unity/Assets/CarriageNode.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

struct State {
    public State(float _fractionalPosition, LineController _lineOccupied, int _direction){
        fractionalPosition = _fractionalPosition;
        lineOccupied = _lineOccupied;
        direction = _direction;
    }

    public float fractionalPosition;
    public LineController lineOccupied;
    public int direction;
}

public class CarriageNode : MonoBehaviour
{
    private LineController lineOccupied;
    private float fractionalPosition;
    private int direction;
    private bool isEndsNode;

    private State previousState;

    private CarriageNode toFollow; // Temporary

    public void Setup(LineController lineOccupied, float fractionalPosition, CarriageNode tf){
        this.lineOccupied = lineOccupied;
        this.fractionalPosition = fractionalPosition;
        this.direction = 1; // 1 if the node is traversing segment start -> finish, else -1 or 0
        this.isEndsNode = false;
        this.toFollow = tf; // Temporary
    }

    void SaveState(){
        previousState = new State(fractionalPosition, lineOccupied, direction);
    }

    void UndoStep(){
        fractionalPosition = previousState.fractionalPosition;
        lineOccupied       = previousState.lineOccupied;
        direction          = previousState.direction;
    }

    int Step(float velocity, bool save=false){
        // Move forward a fixed distance. Returns True on any movement. //
        // Error codes: 0 - success, 1 - failure //
        if (save){ SaveState(); }

        velocity *= direction; // The velocity in the direction in which the train is treating this segment of track
        float lineLength = lineOccupied.length;
        float fractionalChange = velocity / lineLength;
        float newFractionalPosition = fractionalPosition + fractionalChange;
        JunctionController junction;
        float overshoot;

        if (newFractionalPosition >= 0 && newFractionalPosition <= 1){
            // Staying on same line
            fractionalPosition = newFractionalPosition;
            return 0;
        }else {
            if (newFractionalPosition < 0){
                junction = lineOccupied.startJunction;
                overshoot = 0 - newFractionalPosition;
            }else {
                junction = lineOccupied.endJunction;
                overshoot = newFractionalPosition - 1;
            }
        }

        // Switch onto new track:
        if ( !(junction is TriPointsController)){
            Debug.Log("Hit unhandled junction type, stopping. Junction type is: ");
            Debug.Log(junction);
            return 1; // Unhandled junction type, train terminates
        }

        //TriPointsController junction = (TriPointsController) junction;

        LineController nextline = ((TriPointsController) junction).nextLine(lineOccupied);
        Debug.Log("Nextline is: "+nextline);
        if (nextline == null){ // Train can't go this way!
            return 1;
        }

        Debug.Log("Node at "+ this +" steps over junction"+ junction +".");
        if (isEndsNode){
            junction.ToggleOccupied();
        }
        overshoot *= lineLength; // Gives overshoot as an absolute distance
        float fractionalOvershoot = overshoot / nextline.length; // Overshoot as a fraction of the new line

        bool nextlineStartsHere = (nextline.startJunction == junction);
        bool thislineStartsHere = (lineOccupied.startJunction == junction);

        if (nextlineStartsHere){
            // Python code sets previous fracitonalPosition to 0 here.
            fractionalPosition = 0 + fractionalOvershoot;
        }else {
            // Python code sets previous fracitonalPosition to 1 here.
            fractionalPosition = 1 - fractionalOvershoot;
        }

        // If tracks are either nose-nose or tail-tail: flip direction; otherwise, do nothing:
        if ( !(nextlineStartsHere ^ thislineStartsHere) ){
            direction *= -1;
        }

        lineOccupied = nextline;
        return 0;
    }

    int Follow(CarriageNode aheadCarriageNode, int carriageLength, float trainVelocity) {
        // Error codes: 0 - success, 1 - failure
        float relativeTrainVelocity = trainVelocity; // Initial value, independent of direction
        trainVelocity *= direction; // The velocity in the direction in which the train is treating this segment of track (absolute velocity on segment)
        // Find all points which are a carriage length away from aheadCarriageNode on this line
        List<float> intersections = lineOccupied.IntersectsCircle(aheadCarriageNode.GetPosition(), carriageLength);

        if (trainVelocity == 0){
            return 0;
        }

        // Filter out positions "behind" the current position, or ahead of ahead_carriage_node
        List<float> candidatePositions = new List<float>();
        for (int i=0; i<intersections.Count; i++){
            float intersection = intersections[i];
            // Check intersection is ahead of current position
            if (aheadOf(intersection, fractionalPosition, trainVelocity)){
                if (lineOccupied == aheadCarriageNode.lineOccupied){
                    // Then the next step must be behind ahead_carriage_node
                    if (! aheadOf(intersection, aheadCarriageNode.fractionalPosition,trainVelocity)){
                        candidatePositions.Add(intersection);
                    }
                } else {
                    // Not on the same lines, so relative fractionalPositions don't matter
                    candidatePositions.Add(intersection);
                }
            }
        }

        if (candidatePositions.Count == 1){
            // Only one place to move to
            fractionalPosition = candidatePositions[0];
        }else if(candidatePositions.Count == 0){
            if (lineOccupied == aheadCarriageNode.lineOccupied){
                // On same track as node in front, but nowhere to go, so can't move
                return 1;
            }else {
                // Nowhere to go on this track, but there may be somewhere on next track
                if (trainVelocity > 0){
                    // Put this node just beyond junction
                    fractionalPosition =  1.0001f;
                }else {
                    fractionalPosition = -0.0001f;
                }
                LineController initialLineOccupied = lineOccupied;
                //Step(0); // Stepping will now switch this node onto the next line
                Step(0);
                if (lineOccupied == initialLineOccupied){
                    throw new Exception("No valid position for following node at "+this);
                }else {
                    // Try following aheadCarriageNode, now this node is on the next line
                    Debug.Log("Attempting recursive follow!");
                    Follow(aheadCarriageNode, carriageLength, relativeTrainVelocity);
                }
            }
        }else {
            throw new Exception("More than 1 position available to move following node at "+ this +": "+ candidatePositions);
        }

        return 0;
    }

    private bool aheadOf(float carriageNode1, float carriageNode2, float trainVelocity) {
        if (trainVelocity > 0){
            return (carriageNode1 > carriageNode2);
        }else {
            return (carriageNode1 < carriageNode2);
        }
    }

    void MoveTo(CarriageNode aheadCarriageNode) {
        lineOccupied       = aheadCarriageNode.lineOccupied;
        fractionalPosition = aheadCarriageNode.fractionalPosition;
        direction          = aheadCarriageNode.direction;
    }

    Vector2 GetPosition() {
        // Interpolate position of node //
        return lineOccupied.InterpolatePosition(fractionalPosition);
    }

    public override string ToString() {
        // Display position of node //
        return GetPosition()+" on line "+lineOccupied;
    }

    // Start is called before the first frame update
    void Start()
    {

    }

    // Update is called once per frame
    void Update()
    {
        float x = Input.GetAxis("Horizontal");
        if (toFollow == null){
            Step(x*0.1f);
        }else {
            Follow(toFollow,2,x*0.1f);
        }
        Vector2 pos = GetPosition();
        gameObject.transform.position = new Vector3(pos.x,pos.y,1);
    }
}
